### PR TITLE
adding in new loss, need to add in run_experiment pipeline still

### PR DESCRIPTION
**Pull request by Jesse-mk**
adding in new loss

### CODE REVIEW

File: src/run_experiment.py [hunk 1]
```diff
@@ -263,24 +273,32 @@ def main(config: ExperimentConfig):
...
         end = time.time()
         duration = str(timedelta(seconds=end - start)).split('.')[0]
 
         print(
             f'Epoch {epoch}: {train_loss=:.4f}, {valid_loss=:.4f}, {valid_acc=:.4f}, {test_loss=:.4f}, {test_acc=:.4f}. Took {duration}s.'
-        )
+        )        
+#         print(
+#            f'Epoch {epoch}: {train_loss}, {valid_loss}, {valid_acc}, {test_loss}, {test_acc}. Took {duration}s.'
+#        )
```
Comment (Jamin-Chen) [inline src/run_experiment.py:301]: delete

File: src/run_experiment.py [hunk 2]
```diff
@@ -263,24 +273,32 @@ def main(config: ExperimentConfig):
...
             model,
             test_dataloader,
+            loss_weight,
+            loss_c,
             pbar_total=n_test_pages,
             pbar_step=megabatch_size,
             pbar_desc='Test Pages',
         )
 
+
```
Comment (Jamin-Chen) [inline src/run_experiment.py:292]: delete

File: src/data/NLPLLoss.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    min_values = torch.log(((torch.e ** (C* similarities)).sum(axis=-1) * incorrect).sum(axis=1)).sum()
+
+    #combine terms
+    final_loss = (max_values + min_values) * (1/n_total_comparisons) * (1/bs)
+    
+    return final_loss
+
+#get cosine similarities
+
+class NLPLLoss(nn.Module):
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:31]: Can we rename to `CEAndInfoCELoss`?

File: src/data/NLPLLoss.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+from torch.nn.modules.loss import _Loss
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+def calculate_similarities(answer_emb, context_emb, labels, C, bs, device="cuda"):
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:7]:
```suggestion
def calculate_similarities(answer_emb, context_emb, labels, C, batch_size, device='cuda'):
```

File: src/data/NLPLLoss.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        
+        self.C = C
+        self.weight = weight
+        self.device = device
+
+    def forward(self, logits, labels, context_emb, answer_emb):
+        bs = labels.shape[0]
+        
+        #losses
+        primary_loss = F.cross_entropy(logits, labels)
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:43]:
```suggestion
        ce_loss = F.cross_entropy(logits, labels)
```

File: src/data/NLPLLoss.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        print(primary_loss)
+        
+        aux_loss = calculate_similarities(answer_emb, context_emb, labels, self.C, bs, self.device)
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:46]:
```suggestion
        info_ce_loss = calculate_similarities(answer_emb, context_emb, labels, self.C, bs, self.device)
```

File: src/data/NLPLLoss.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    #combine terms
+    final_loss = (max_values + min_values) * (1/n_total_comparisons) * (1/bs)
+    
+    return final_loss
+
+#get cosine similarities
+
+class NLPLLoss(nn.Module):
+    def __init__(self, C=0.1, weight=0.5, device="cuda"):
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:32]:
```suggestion
    def __init__(self, C=0.1, lambda=0.5, device="cuda"):
```

In paper I think we will refer to this parameter as lambda so its best just to have the code match.

File: src/data/NLPLLoss.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    def forward(self, logits, labels, context_emb, answer_emb):
+        bs = labels.shape[0]
+        
+        #losses
+        primary_loss = F.cross_entropy(logits, labels)
+        print(primary_loss)
+        
+        aux_loss = calculate_similarities(answer_emb, context_emb, labels, self.C, bs, self.device)
+        
+        print(aux_loss)
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:48]: delete

File: src/data/NLPLLoss.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        self.C = C
+        self.weight = weight
+        self.device = device
+
+    def forward(self, logits, labels, context_emb, answer_emb):
+        bs = labels.shape[0]
+        
+        #losses
+        primary_loss = F.cross_entropy(logits, labels)
+        print(primary_loss)
```
Comment (Jamin-Chen) [inline src/data/NLPLLoss.py:44]: delete

**Review by Jamin-Chen — CHANGES_REQUESTED** [review]
Can you install Black code formatter (`pip install black`) and then run with the following command in the root directory ?
```
black . --skip-string-normalization
```
